### PR TITLE
DXF Deserialize

### PR DESCRIPTION
**Pull request by z3dev**
WORK IN PROGRESS DO NOT MERGE

TASKS
- [X] Review DXF specifications
- [X] Review DXF libraries
- [X] Obtain samples for 2D and 3D test suites
- [X] Implement DXF reader class
- [X] Implement 2D import to objects (instantiate)
- [X] Complete test suite for 2D instantiate
- [X] Implement 3D import to objects (instantiate)
- [X] Complete test suite for 3D instantiate
- [X] Implement 2D translation to CAG source (translate)
- [X] Complete test suite for 2D translate
- [X] Implement 3D translation to CSG source (translate)
- [X] Complete test suite for 3D translate
- [X] Complete JSDOC comments
- [X] Complete README

### CODE REVIEW

**Comment by kaosat-dev** [comment]
Wow , you have been busy ! 
Great work @z3dev !

**Comment by z3dev** [comment]
OK. This is ready for review.
DXF is a nasty old format, and this is a good starting place for supporting both 2D and 3D entities from Autodesk.

File: README.md [hunk 1]
```diff
@@ -11,11 +11,22 @@ This repository includes all the input/output format handling for the JSCAD proj
...
+- [x] amf-deserializer
+- [x] dxf-deserializer
+- [ ] json-deserializer
+- [x] obj-deserializer
+- [x] stl-deserializer
+- [x] svg-deserializer
 
 Following formats are supported as inputs:
  - [AMF](https://github.com/jscad/io/blob/master/packages/amf-deserializer)
+ - [DXF](https://github.com/jscad/io/blob/master/packages/dxf-deserializer)
```
Comment (kaosat-dev) [inline README.md]: Nice that you did not forget adding these to the main README :)

Comment (kaosat-dev) [inline README.md]: Nice that you did not forget adding these to the main README :)

File: packages/amf-deserializer/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# Change Log
```
Comment (kaosat-dev) [inline packages/amf-deserializer/CHANGELOG.md]: These are from different PRs/commits , we need to rebase

File: packages/dxf-deserializer/test/test-2d-entities.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
+const fs = require('fs')
+const path = require('path')
+const test = require('ava')
+const { CSG, CAG } = require('@jscad/csg')
+
+import {nearlyEqual} from '../../../test/helpers/nearlyEqual'
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/test/test-2d-entities.js]: this is an es6 import , to be avoided, please use require()

File: packages/dxf-deserializer/package.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  "keywords": [
+    "openjscad",
+    "jscad",
+    "csg",
+    "deserializer",
+    "dxf"
+  ],
+  "license": "MIT",
+  "dependencies": {
+    "@jscad/csg": "^0.3.7"
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/package.json]: The version of csg.js to use is now 0.5.2 , but perhaps that should be done in bulk in another PR

File: packages/dxf-deserializer/test/test-2d-translation.js [hunk 5]
```diff
@@ -0,0 +1,346 @@
+const fs = require('fs')
+const path = require('path')
+const test = require('ava')
+const { CSG, CAG } = require('@jscad/csg')
+
+import {nearlyEqual} from '../../../test/helpers/nearlyEqual'
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/test/test-2d-translation.js]: same as in the other tests, needs to use require()

File: packages/dxf-deserializer/test/test-3d-entities.js [hunk 6]
```diff
@@ -0,0 +1,176 @@
+const fs = require('fs')
+const path = require('path')
+const test = require('ava')
+const { CSG, CAG } = require('@jscad/csg')
+
+import {nearlyEqual} from '../../../test/helpers/nearlyEqual'
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/test/test-3d-entities.js]: same

File: packages/dxf-deserializer/test/test-3d-entities.js [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+  t.true(Array.isArray(objs))
+  t.is(objs.length,5)
+
+  let obj4 = objs[4]
+  t.true(obj4 instanceof CSG)
+  t.is(obj4.toPolygons().length,6)
+})
+
+
+test('ASCII DXF from Autocad2017 3D Mesh to Object Conversion',  t => {
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/test/test-3d-entities.js]: lots of very nice tests ! 👍

File: packages/dxf-deserializer/test/test-dxf-versions.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
+const fs = require('fs')
+const path = require('path')
+const test = require('ava')
+const { CSG, CAG } = require('@jscad/csg')
+
+import {nearlyEqual} from '../../../test/helpers/nearlyEqual'
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/test/test-dxf-versions.js]: require() instead of import

File: packages/dxf-deserializer/test/test-dxf-versions.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+const { deserialize } = require( '../index' )
+
+const samples = path.resolve('../../node_modules/@jscad/sample-files')
+
+//
+// Test suite for DXF deserialization (import)
+//
+test('ASCII DXF R13 to Object Conversion', t => {
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/test/test-dxf-versions.js]: eek , I always forget how many different versions of the format there are, this is huge work , kudos !

File: packages/dxf-deserializer/translate.js [hunk 10]
```diff
@@ -0,0 +1,880 @@
...
+        translatePath2D(obj, layers, options)
+        break
+
+      default:
+        console.log('##### ERROR')
+        console.log(obj.type)
+        break
+    }
+    // accumlate polygons if necessary
+    if (p instanceof CSG.Polygon) {
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/translate.js]: two things : 
- could the different p instanceof XXX not be combined ?
- as with 'instanceof CSG/CAG' perhaps it might be better to test for object properties with some helpers ? (there is always a risk of failure to identify types) : I have a few solutions for that going forward , with changes to CSG.js so this might need to wait ?

File: packages/dxf-deserializer/translate.js [hunk 11]
```diff
@@ -0,0 +1,880 @@
...
+      // 2D entities
+      case 'lwpolyline':
+        console.log('##### lwpolyline')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        translatePath2D(obj, layers, options)
+        break
+
+      default:
+        console.log('##### ERROR')
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/translate.js]: perhaps we should add a debug flag to the serializers/deserializers or a logging system to avoid ending up with console messages in production

File: packages/dxf-deserializer/translate.js [hunk 12]
```diff
@@ -0,0 +1,880 @@
...
+  layers.forEach(
+    function (layer) {
+      let name = layer['name'] || 'Unknown'
+      script += 'layer' + name + '(),'
+    }
+  )
+  script += '[])\n}\n'
+
+  // add helper functions for polygons and lines
+  script += 'function createVertex(point) {\n'
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/translate.js]: it would be better to use es6 string templates (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) instead of string concatenation whenever possible

File: packages/dxf-deserializer/DxfReader.js [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+
+All code released under MIT license
+
+Thanks to @issacs for the sax js library, and inspiration for this reader
+
+*/
+
+// DXF reader to emit groups of interest to handlers
+
+;(function (dxf) { // wrapper for non-node envs
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/DxfReader.js]: this is strange: any reason why this is not structured as a 'normal' module ?

File: packages/dxf-deserializer/index.js [hunk 14]
```diff
@@ -0,0 +1,592 @@
...
+const translateAsciiDxf = require('./translate')
+
+// //////////////////////////////////////////
+//
+// DXF (Drawing Exchange Format) is a CAD data file format developed by Autodesk
+//
+// //////////////////////////////////////////
+
+function handleError (e) {
+  console.log('error: line ' + e.line + ', column ' + e.column + ', bad character [' + e.c + ']')
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/index.js]: Errors should throw like for the other de/serializers , otherwise there is no way to handle the errors in the rest of the app or propagate them.
Same note as before for console.log output

File: packages/dxf-deserializer/index.js [hunk 15]
```diff
@@ -0,0 +1,592 @@
...
+function handleEnd (reader, data) {
+  console.log('DXF reader completed')
+}
+
+//
+// handle a entity as provided by the reader
+// groups: 0
+// special handling to set defaults as per DXF specifications
+//
+function handleEntity (reader, group, value) {
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/index.js]: insane amount of work ! I cannot even imagine the headaches to deal with the DXF specs!

File: packages/dxf-deserializer/instantiate.js [hunk 16]
```diff
@@ -0,0 +1,584 @@
+/*
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/instantiate.js]: so much work ! :)

File: packages/dxf-deserializer/DxfReader.js (hunk 13 above)
Comment (z3dev) [reply packages/dxf-deserializer/DxfReader.js]: Taken from the SAX parser. I’ll change to normal exports..

File: packages/amf-deserializer/CHANGELOG.md (hunk 2 above)
Comment (z3dev) [reply packages/amf-deserializer/CHANGELOG.md]: Is this automatic as part of lerna? Would be nice..

File: packages/dxf-deserializer/test/test-2d-entities.js (hunk 3 above)
Comment (z3dev) [reply packages/dxf-deserializer/test/test-2d-entities.js]: Ok. Will change...

File: packages/dxf-deserializer/package.json (hunk 4 above)
Comment (z3dev) [reply packages/dxf-deserializer/package.json]: I’ll update and test. Should be fine.

File: packages/dxf-deserializer/test/test-dxf-versions.js (hunk 9 above)
Comment (z3dev) [reply packages/dxf-deserializer/test/test-dxf-versions.js]: As many as there are years from 1901... ha ha ha

File: packages/dxf-deserializer/translate.js (hunk 11 above)
Comment (z3dev) [reply packages/dxf-deserializer/translate.js]: Good idea.  The strict option exists so I’ll change logging and exception throwing to check this option.

File: packages/amf-deserializer/CHANGELOG.md (hunk 2 above)
Comment (kaosat-dev) [reply packages/amf-deserializer/CHANGELOG.md]: Lerna creates changelogs automatically, but you have commits from a different context, that has nothing to do with Lerna: )

File: packages/dxf-deserializer/instantiate.js (hunk 16 above)
Comment (kaosat-dev) [reply packages/dxf-deserializer/instantiate.js]: are all those versions official supported still ?

Comment (z3dev) [reply packages/dxf-deserializer/instantiate.js]: It’s not so much supported but the range of applications creating DXF files is staggering. A robust implementation shouldn’t die, so I added simple tests.

File: packages/dxf-deserializer/translate.js (hunk 12 above)
Comment (z3dev) [reply packages/dxf-deserializer/translate.js]: Seems that MS Explorer doesn’t implement templates. Maybe it’s still too early to switch. Thoughts?

File: packages/dxf-deserializer/test/test-2d-entities.js (hunk 3 above)
Comment (z3dev) [reply packages/dxf-deserializer/test/test-2d-entities.js]: Changes made.

File: packages/dxf-deserializer/package.json (hunk 4 above)
Comment (z3dev) [reply packages/dxf-deserializer/package.json]: Changes made.

File: packages/dxf-deserializer/translate.js (hunk 10 above)
Comment (z3dev) [reply packages/dxf-deserializer/translate.js]: Since these functions are PRODUCING CSG / CAG / etc objects then the instanceof() methods are valid, i.e. everything being produced is 'correct'.

File: packages/dxf-deserializer/DxfReader.js (hunk 13 above)
Comment (z3dev) [reply packages/dxf-deserializer/DxfReader.js]: Hmm... I stumped. I can make dxf into an object. Or I can export DxfReader. Ideally, dxf is just a name space. Suggestions?

File: packages/dxf-deserializer/index.js (hunk 14 above)
Comment (z3dev) [reply packages/dxf-deserializer/index.js]: Changes made.
All logging of debug statements are commented out.
Error handling revised.

File: packages/dxf-deserializer/DxfReader.js (hunk 13 above)
Comment (kaosat-dev) [reply packages/dxf-deserializer/DxfReader.js]: Hmm if it works like this , just leave it be for now, I'll take a closer look : I already did similar changes for AMF & SVG so we should align with that, but it is something for a next version I think :)

**Comment by z3dev** [comment]
Mark, if done with the review, then I could put these changes into a new PR as a single commit. Thoughts?

File: packages/amf-deserializer/CHANGELOG.md (hunk 2 above)
Comment (z3dev) [reply packages/amf-deserializer/CHANGELOG.md]: I checked the differences between this PR and the master branch. The only other changes are to the README.md files in packages/io, and the addition of test/helpers/nearlyEqual.js

Is there something else?

**Comment by z3dev** [comment]
Maybe the PR should be merged using "Rebase and merge". I think that may make the merge cleaner.

File: packages/dxf-deserializer/translate.js (hunk 12 above)
Comment (kaosat-dev) [reply packages/dxf-deserializer/translate.js]: we have loooong since given up support for IE(failing webgl support in there in the first place), Edge supports them : remember : the past is the past :)

File: packages/amf-deserializer/CHANGELOG.md (hunk 2 above)
Comment (kaosat-dev) [reply packages/amf-deserializer/CHANGELOG.md]: I see a lot of changes in other packages as well, I will try to do a rebase locally to see if that would fix things

File: packages/dxf-deserializer/translate.js (hunk 12 above)
Comment (z3dev) [reply packages/dxf-deserializer/translate.js]: OK! We e given ya all there is captain! Warp 10!

Comment (z3dev) [reply packages/dxf-deserializer/translate.js]: I’ll make the changes tomorrow. :)

File: packages/amf-deserializer/CHANGELOG.md (hunk 2 above)
Comment (z3dev) [reply packages/amf-deserializer/CHANGELOG.md]: Any chance that you have a old copy of IO master? Hmm...

**Comment by z3dev** [comment]
I enhanced the code to use string templates where possible. Please review.

File: packages/amf-deserializer/CHANGELOG.md (hunk 2 above)
Comment (kaosat-dev) [reply packages/amf-deserializer/CHANGELOG.md]: I have an up to date master, so that is all we need :) this should be saved for last :)

File: packages/dxf-deserializer/index.js [hunk 17]
```diff
@@ -575,7 +578,7 @@ const deserialize = function (src, filename, options) {
   const defaults = {
     version: '0.0.1',
     output: 'jscad',
-    strict: true,
+    strict: false,
```
Comment (kaosat-dev) [inline packages/dxf-deserializer/index.js]: why the change to false ? Sorry If I am annoying with this, but parsers should ALWAYS throw : how that error gets propagated and handled is the job of the consumer (web/desktop app/cli /else), without an exception, there is no way of knowing what happened inside the de/serializers, and the app(s) could enter an unknown state

**Comment by kaosat-dev** [comment]
All the string templating looks good to me !

**Comment by kaosat-dev** [comment]
copy and pasting this back here since the comment seems hidden: 
why the change to false for `strict` ? Sorry If I am annoying with this, but parsers should ALWAYS throw : how that error gets propagated and handled is the job of the consumer (web/desktop app/cli /else), without an exception, there is no way of knowing what happened inside the de/serializers, and the app(s) could enter an unknown state

**Comment by z3dev** [comment]
For one reason, a previous comment. For another reason, the use of strict was inconsistent, which should be fixed now.

So, it’s just a matter of setting the option to true now.

**Comment by kaosat-dev** [comment]
The other version of this PR has been merged, so closing this one !
Thanks for the awesome, hard work Jeff ! :)
